prov-api/helpers/generate_report_helpers.py

```diff
@@ -0,0 +1,223 @@
+
+    if isinstance(response, str):
+        raise HTTPException(status_code=400, detail=f"Validation error with provided {item_subtype.value}: {response}")
+    
+    if isinstance(response, SeededItem):
+        raise HTTPException(status_code=400, detail="Seeded item cannot be used for this query!")
+    
+
+def filter_for_export_prov_graph(upstream_nodes: List[Node], downstream_nodes: List[Node]) -> Dict[str,List[Node]]: 
+    """Takes a list of upstream and downstream nodes and processes them into into different categories of 
+    inputs, models and outputs. 
+
+    Parameters
+    ----------
+    upstream_nodes : List[Node]
+        A list of all upstream nodes. 
+    downstream_nodes : List[Node]
+        A list of all downstream nodes.
+
+    Returns
+    -------
+    Dict[str,List[Node]]
+        A dictionary containing the "inputs", "outputs" and "model-runs" involved within the study/model-run.
+    """
+
+    # Filter through both the upstream and downstream nodes. 
+    filtered_response: Dict[str,List[Node]] = {
+        "inputs": [], 
+        "model_runs": [], 
+        "outputs": []
+    }
+
+    # Small helper function to process the node and add it accordingly. 
+    def process_node(node: Node, node_type: str) -> None: 
+        if node.item_subtype in [ItemSubType.MODEL, ItemSubType.DATASET]:
+            filtered_response[node_type].append(node)
+        elif node.item_subtype == ItemSubType.MODEL_RUN:
+            filtered_response["model_runs"].append(node)
+    
+    for node in upstream_nodes:
+        process_node(node, "inputs")
+
+    for node in downstream_nodes: 
+        process_node(node, "outputs")
+
+    return filtered_response
+
+def parse_nodes(node_list: List[Any]) -> List[Node]: 
+    """Parses a list of potential nodes into a list of Valid Node objects.
+
+    This function takes a list of potential nodes, attempts to parse them into 
+    the Node Pydantic model, and returns a list of validated Node objects. 
+
+    Parameters
+    ----------
+    node_list : List[Any]
+        A list of items that can potentially be parsed as Node objects.
+
+    Returns
+    -------
+    List[Node]
+        A list of parsed Node objects.
+
+
+    Raises
+    ------
+    HTTPException
+        Raised if the provided node_list is empty.
+    HTTPException
+        Raised if validation fails while parsing any node in the list.
+    """
+
+    node_list_parsed: List[Node] = []
+
+    if len(node_list) == 0: 
+        # Do I handle this? Can there be a case where a model run does not have any assoicated inputs/outputs?? - I don't think so.
+        raise HTTPException(status_code=400, detail="Cannot process the report generation.")
+    try:
+        for node in node_list: 
+            node_list_parsed.append(Node(**node))
+    except ValidationError as e: 
+        raise HTTPException(status_code=400, detail="Something has gone wrong in parsing this node.")
```

Comment:
and also "this node" is not specific. Maybe include "... in parsing node of ID {node.id}

Reply:
Okay I have structured the try and catch block to capture more specific details of the node. Probably not the best idea to do node.id - since it has not been parsed yet, however I did node.get('id', 'undefined id') to be safe.